### PR TITLE
Tracking for setting attributes

### PR DESCRIPTION
**Pull request by dgasmith**
## Change Summary

Tracking when default arguments are override and allows a `skip_defaults` kwarg in the `dict` method. 

A problem that I have not yet solved is that a BaseModel goes through the `dict_validator` and is then passed back into `BaseModel.validate` after being cast to a dict, this does not allow the `__fields_set__` to come through at the moment and recursive models do not work. Any suggestions here?

```
make benchmark-pydantic
Before:           pydantic best=29.726μs/iter avg=30.439μs/iter stdev=0.563μs/iter
After:            pydantic best=29.061μs/iter avg=29.695μs/iter stdev=0.473μs/iter
```
## Related issue number

Discussion in #378.

## Checklist

* [x] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [ ] Documentation reflects the changes where applicable
* [ ] `HISTORY.rst` has been updated
  * if this is the first change since a release, please add a new section
  * include the issue number or this pull request number `#<number>`
  * include your github username `@<whomever>`

### CODE REVIEW

**Comment by samuelcolvin** [comment]
Single quotes please everywhere, also you'll need to resplect black. Use `make format`.

**Comment by dgasmith** [comment]
Of course, I was holding off on that area until we had an idea of what to do with the recursive copy operation. Easiest thing would be to write a custom copy validator for BaseModels rather than relying on the standard `dict` validator. However, that solution isn't as elegant as the current code.

**Comment by samuelcolvin** [comment]
Oh I see. Maybe related to #265?

Do you know how to proceed or do you have a specific question?

**Comment by dgasmith** [comment]
Similar to #265. If you are fine with having a custom validator for BaseModels rather than the `dict_validator` we could solve both the issue here and #265.

**Comment by samuelcolvin** [comment]
great

**Comment by dgasmith** [comment]
#265 was deeper than I expected. I think that can be solved however, but I will leave it for another PR. This is now ready to review.

File: pydantic/main.py [hunk 1]
```diff
@@ -206,12 +206,14 @@ class BaseModel(metaclass=MetaModel):
...
     Config = BaseConfig
-    __slots__ = ('__values__',)
+    __slots__ = ('__values__', '__fields_set__')
 
     def __init__(self, **data: Any) -> None:
         if TYPE_CHECKING:  # pragma: no cover
             self.__values__: Dict[str, Any] = {}
-        self.__setstate__(self._process_values(data))
+            self.__fields_set__: Set[str] = set()
+        fields_set = data.pop("__fields_set__", set(data.keys()))
```
Comment (samuelcolvin) [inline pydantic/main.py]: Strict rule in pydantic that you can;t pass anything other than values to the class like this.

Users should be able to create a model with `MyModel(**whatever)` without having to think about what's in `whatever`.

Remove this and either call `__setstate__` directly, or set `model.__fields_set__ = thing` directly.

File: pydantic/main.py [hunk 2]
```diff
@@ -232,26 +234,37 @@ def __setattr__(self, name, value):
                 raise ValidationError([error_])
             else:
                 self.__values__[name] = value_
+                self.__fields_set__.update({name})
```
Comment (samuelcolvin) [inline pydantic/main.py]:
```suggestion
                self.__fields_set__.add(name)
```

File: pydantic/main.py [hunk 3]
```diff
@@ -232,26 +234,37 @@ def __setattr__(self, name, value):
...
         else:
             self.__values__[name] = value
+            self.__fields_set__.update({name})
```
Comment (samuelcolvin) [inline pydantic/main.py]:
```suggestion
            self.__fields_set__.add(name)
```

File: pydantic/main.py [hunk 4]
```diff
@@ -324,7 +340,8 @@ def construct(cls, **values: Any) -> 'BaseModel':
         Chances are you don't want to use this method directly.
         """
         m = cls.__new__(cls)
-        m.__setstate__(values)
+        fields_set = values.pop("__fields_set__", None)
```
Comment (samuelcolvin) [inline pydantic/main.py]: Single quotes always please unless the string contains single quotes.

But again, can we avoid doing `values.pop()`.

File: pydantic/main.py [hunk 5]
```diff
@@ -342,15 +359,13 @@ def copy(
         """
         if include is None and exclude is None and update is None:
             # skip constructing values if no arguments are passed
-            v = self.__values__
+            v = self.__values__.copy()
```
Comment (samuelcolvin) [inline pydantic/main.py]: do we need this copy?

File: pydantic/main.py [hunk 6]
```diff
@@ -418,9 +433,25 @@ def __iter__(self) -> 'AnyGenerator':
...
+    def _iter(self, by_alias: bool = False, skip_defaults: bool = True) -> 'TupleGenerator':
         for k, v in self.__values__.items():
-            yield k, self._get_value(v, by_alias=by_alias)
+            yield k, self._get_value(v, by_alias=by_alias, skip_defaults=skip_defaults)
+
+    def _calculate_keys(
+        self, include: Set[str] = None, exclude: Optional[Set[str]] = set(), skip_defaults: bool = False
+    ) -> Set[str]:
+        if skip_defaults:
+            keys = self.__fields_set__.copy()
```
Comment (samuelcolvin) [inline pydantic/main.py]: i doubt we need `copy()` here.

File: pydantic/main.py [hunk 7]
```diff
@@ -418,9 +433,25 @@ def __iter__(self) -> 'AnyGenerator':
...
+        else:
+            keys = set(self.__values__.keys())
```
Comment (samuelcolvin) [inline pydantic/main.py]: I doubt we need `set()` here.

File: pydantic/main.py [hunk 8]
```diff
@@ -418,9 +433,25 @@ def __iter__(self) -> 'AnyGenerator':
...
+
+        if include:
+            keys &= set(include)
```
Comment (samuelcolvin) [inline pydantic/main.py]: I doubt we need `set()` here.

File: pydantic/main.py [hunk 9]
```diff
@@ -418,9 +433,25 @@ def __iter__(self) -> 'AnyGenerator':
         """
         yield from self._iter()
 
-    def _iter(self, by_alias: bool = False) -> 'TupleGenerator':
+    def _iter(self, by_alias: bool = False, skip_defaults: bool = True) -> 'TupleGenerator':
         for k, v in self.__values__.items():
-            yield k, self._get_value(v, by_alias=by_alias)
+            yield k, self._get_value(v, by_alias=by_alias, skip_defaults=skip_defaults)
+
+    def _calculate_keys(
```
Comment (samuelcolvin) [inline pydantic/main.py]: I think we can probably do this in a more efficient way, eg. if `include=None`, `exclude=set()` and `skip_defaults=False` (all the defaults), then we just take all keys.

Can't we take care of that case but just returning `None` or something here and not checking key containment?

File: HISTORY.rst [hunk 10]
```diff
@@ -3,6 +3,9 @@
 History
 -------
 
+* Adds ``skip_defaults`` argument to ``BaseModel.dict()`` to allow skipping of fields that were not
```
Comment (samuelcolvin) [inline HISTORY.rst]:
```suggestion
v0.20.0 (unreleased)
....................
* Adds ``skip_defaults`` argument to ``BaseModel.dict()`` to allow skipping of fields that were not
```

File: pydantic/main.py (hunk 6 above)
Comment (dgasmith) [reply pydantic/main.py]: We do as we modify the set in-place later for speed.

File: pydantic/main.py (hunk 7 above)
Comment (dgasmith) [reply pydantic/main.py]: If we do not convert it to a set (there is no copy here). We rely on the fact that dict keys are special sets which return LHS references if modified in place. Seems a bit ugly, when we can explicitly copy the data here and move on with things.

File: pydantic/main.py (hunk 8 above)
Comment (dgasmith) [reply pydantic/main.py]: I was erring on the side of safety, not everyone uses mypy. But happy to remove it.

File: pydantic/main.py (hunk 9 above)
Comment (dgasmith) [reply pydantic/main.py]: Sure, I can work on optimizing that out.

File: pydantic/main.py (hunk 5 above)
Comment (dgasmith) [reply pydantic/main.py]: No, not if we remove the explicit add of `__field_set__` below.

File: pydantic/main.py (hunk 8 above)
Comment (samuelcolvin) [reply pydantic/main.py]: I think this one really can be removed. If someone passes something weird here they should get an error.

File: pydantic/main.py (hunk 7 above)
Comment (samuelcolvin) [reply pydantic/main.py]: ok, that's fine.

File: pydantic/main.py (hunk 6 above)
Comment (samuelcolvin) [reply pydantic/main.py]: ok

File: pydantic/main.py (hunk 9 above)
Comment (dgasmith) [reply pydantic/main.py]: On this note, the include/exclude defaults are not consistent (copy for example has them both defaulting to None). I can clean this up as well.

Comment (samuelcolvin) [reply pydantic/main.py]: ok.

File: pydantic/main.py (hunk 4 above)
Comment (dgasmith) [reply pydantic/main.py]: I think I need to add the ability to pass in some sort of state from the calling instance. Can I change `construct` to take in a `fields_set` or should I add a new function like `construct_with_fields` or similar?

Comment (samuelcolvin) [reply pydantic/main.py]: `construct` can take more arguments, just `__init__` that should remain pure.

**Comment by dgasmith** [comment]
Timings:
```
defaultfields  BENCHMARK_REPEATS=20 make benchmark-pydantic
python benchmarks/run.py pydantic-only
                                pydantic time=0.940s, success=48.45%
...
                                pydantic best=0.886s, avg=0.911s, stdev=0.017s

                                pydantic best=7.387μs/iter avg=7.592μs/iter stdev=0.138μs/iter
```
```
master  BENCHMARK_REPEATS=20 make benchmark-pydantic
python benchmarks/run.py pydantic-only
                                pydantic time=0.919s, success=48.45%
...
                                pydantic best=0.888s, avg=0.914s, stdev=0.020s

                                pydantic best=7.397μs/iter avg=7.620μs/iter stdev=0.165μs/iter
```

Not entirely sure why. Could depend heavily on benchmark construction and a bit of extra set manipulation over raw lists.

**Comment by samuelcolvin** [comment]
Tried it myself and I get a ~5% time increase with this PR. Not ideal.

To put this in proportion, this increase is roughly the same as the increase which was initially introduced by the typing PR #373, I spent a couple of hours fighting with it to mostly eliminate that performance deterioration. I'm therefore somewhat reticent about introducing that slowdown here. 

Get the PR ready and I'll see what I can do.

**Comment by dgasmith** [comment]
Odd on timings, those were 3.6. For 3.7 I see:
```
this branch               pydantic best=6.871μs/iter avg=7.165μs/iter stdev=0.169μs/iter
master                    pydantic best=6.841μs/iter avg=7.238μs/iter stdev=0.170μs/iter
```

Agreed on the keys comment, patched.

**Comment by samuelcolvin** [comment]
For me this is now ready to be merged. Please check my changes and let me know if you disagree with anything or have any questions.

We might need to update history to state that `contruct`'s signature has changed.

**Comment by dgasmith** [comment]
Looks good! I was hesitant to change `constructor`, but this is much better and is extensible if there are other slots needed in the future. Thanks for working through this and the great feedback.

**Comment by achaayb** [comment]
Please provide an example usecase.
where should i pass the skip arg

**Pull request merged**